lib/solarwinds_apm/opentelemetry/otlp_processor.rb

```diff
@@ -0,0 +1,169 @@
+# frozen_string_literal: true
+
+# © 2023 SolarWinds Worldwide, LLC. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at:http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+
+module SolarWindsAPM
+  module OpenTelemetry
+    # reference: OpenTelemetry::SDK::Trace::SpanProcessor; inheritance: SolarWindsProcessor
+    class OTLPProcessor < SolarWindsProcessor
+      attr_accessor :description
+
+      # @param [Hash] meters the hash of meter created by ::OpenTelemetry.meter_provider.meter('meter_name')
+      # @param [TxnNameManager] txn_manager storage for transaction name
+      # @exporter [Exporter] exporter reporter that send trace data
+      def initialize(meters, exporter, txn_manager)
+        super(exporter, txn_manager)
+        @meters  = meters
+        @metrics = {}
+        @trace_span_id = nil
+      end
+
+      # Called when a {Span} is started, if the {Span#recording?}
+      # returns true.
+      #
+      # This method is called synchronously on the execution thread, should
+      # not throw or block the execution thread.
+      #
+      # @param [Span] span the {Span} that just started.
+      # @param [Context] parent_context the
+      #  started span.
+      def on_start(span, parent_context)
+        SolarWindsAPM.logger.debug { "[#{self.class}/#{__method__}] processor on_start span: #{span.inspect}, parent_context: #{parent_context.inspect}" }
+
+        initialize_metrics if @metrics.empty?
+
+        @trace_span_id = "#{span.context.hex_trace_id}-#{span.context.hex_span_id}"
+        parent_span    = ::OpenTelemetry::Trace.current_span(parent_context)
+        return if parent_span && parent_span.context != ::OpenTelemetry::Trace::SpanContext::INVALID && parent_span.context.remote? == false
+
+        span_attrs = span_attributes(span)
+        span.add_attributes(span_attrs)
+
+        trace_flags = span.context.trace_flags.sampled? ? '01' : '00'
+        @txn_manager.set_root_context_h(span.context.hex_trace_id, "#{span.context.hex_span_id}-#{trace_flags}") # this is for custom api set_transaction_name to be able to retrieve right span_id from trace_id
```

Comment:
I wonder if the processing in this `on_start` method can all be moved to `on_end`. Or if there must be something done on span start, it seems it should _only_ be for the entry span (i.e. root/no parent, or parent is remote) -- for example it seems wasteful to initialize_metrics on every span.

Reply:
> initialize_metrics on every span.

initialize_metrics will be called if `@metrics` is empty, so metrics only initialize once for the entire cycle of user application.

> span.add_attributes(span_attrs)

Needs to add attributes at on_start since once span.finish (which will call processor.on_finish), the span is finished and doesn't allow adding attributes.

> @txn_manager.set_root_context_h

This will only be called for entry span that is guarded by `return if parent_span && parent_span.context != ::OpenTelemetry::Trace::SpanContext::INVALID && parent_span.context.remote? == false`

Reply:
From discussion, setting the `sw.transaction` span attribute with custom transaction name specified by SDK will not work in span `on_start`.  Let's not support SDK-set custom transaction name for this first iteration (I'll clarify the spec) and remove this need to do anything on span start.